keps/sig-scheduling/5007-device-attach-before-pod-scheduled/README.md

```diff
@@ -0,0 +1,1117 @@
+    //
+    // The maximum number of attributes and capacities combined is 32.
+    //
+    // +optional
+    Attributes map[QualifiedName]DeviceAttribute
+
+    // BindingGates defines the gates for binding.
+    //
+    // +optional
+    BindingGates []string
+
+    // BindingFailureGates defines the gates for binding failure.
+    //
+    // +optional
+    BindingFailureGates []string
+
+    // UsageRestrictedToNode indicates if the usage of an allocation involving this device
+    // has to be limited to exactly the node that was chosen when allocating the claim.
+    //
+    // +optional
+    UsageRestrictedToNode bool
+
+    // BindingTimeout indicates the prepare timeout period(minute).
+    // If the timeout period is exceeded, the scheduler clears the allocation in the ResourceClaim and reschedules the Pod.
+    //
+    // +optional
+    BindingTimeout int
+}
+```
+
+#### AllocatedDeviceStatus Enhancements
+
+The `BindingGates` and `BindingFailureGates` fields within `AllocatedDeviceStatus` are used to indicate the status of the device attachment.
+These fields will contain a list of conditions, each representing a specific state or event related to the device.
+
+For this feature, following fields are added:
+
+```go
+// AllocatedDeviceStatus contains the status of an allocated device, if the
+// driver chooses to report it. This may include driver-specific information.
+type AllocatedDeviceStatus struct {
+    ...
+    // BindingGates defines the gates for binding.
+    //
+    // +optional
+    BindingGates map[string]bool
+
+    // BindingFailureGates defines the gates for binding failure.
+    //
+    // +optional
+    BindingFailureGates map[string]bool
+}
+```
+
+#### Scheduler DRA Plugin Modifications
+
+When `UsageRestrictedToNode: true` is set, the scheduler DRA plugin will perform the following steps:
+
+1. **Set NodeSelector**: Before the `PreBind` phase, add the `NodeName` to the `ResourceClaim`'s `NodeSelector`.
+
+If Gates are present, the scheduler DRA plugin will perform the following steps during the `PreBind` phase:
+
+2. **Copy Gates**: Copy `BindingGates` and `BindingFailureGates` from `ResourceSlice.Device.Basic` to `AllocatedDeviceStatus`.
+3. **Wait for Conditions**: Wait for the following conditions:
+   - Wait until all conditions in the BindingGates are `True` before proceeding to Bind.
+   - If any one of the conditions in the BindingFailureGates becomes `True`, clear the allocation in the `ResourceClaim` and reschedule the Pod.
+   - If the preparation of a device takes longer than the `BindingTimeout` period, clear the allocation in the `ResourceClaim` and reschedule the Pod.
+
+To support these steps, for example, a DRA driver can include the following definitions in BindingGates or BindingFailureGates within a ResourceSlice:
+
+```go
+const (
+    // NeedToPreparing indicates that this device needs some preparation.
+    NeedToPreparing = "kubernetes.io/need-to-preparing"
```

Comment:
I agree with https://github.com/kubernetes/enhancements/pull/5012/files#r1949713686, this is just an example. Because it's specific to a driver, it shouldn't use the `kubernetes.io` domain. You can use `dra.example.com` instead.

Reply:
About "need-to-preparing": I don't see that being used. The binding gate (or rather, condition) is `is-prepared`.